contracts/pcv/PCVGuardian.sol

```diff
@@ -76,114 +79,167 @@ contract PCVGuardian is IPCVGuardian, CoreRef {
         address pcvDeposit,
         address safeAddress,
         address token,
         uint256 amount,
-        bool pauseAfter,
         bool depositAfter
-    ) external override hasAnyOfThreeRoles(TribeRoles.GOVERNOR, TribeRoles.PCV_SAFE_MOVER_ROLE, TribeRoles.GUARDIAN) {
-        require(isSafeAddress(safeAddress), "Provided address is not a safe address!");
+    ) external override beforeAndAfterWithdrawal(pcvDeposit, safeAddress, depositAfter) {
+        IPCVDeposit(pcvDeposit).withdrawERC20(token, safeAddress, amount);
+        emit PCVGuardianERC20Withdrawal(pcvDeposit, safeAddress, token, amount);
+    }
 
-        pcvDeposit._ensureUnpaused();
+    /// @notice withdraw funds from a pcv deposit, by calling the withdrawERC20() method on it
+    /// @param pcvDeposit the deposit to pull funds from
+    /// @param safeAddress the destination address to withdraw to
+    /// @param basisPoints the percent in basis points [1-10000] if the deposit's balance to withdraw
+    /// @param depositAfter if true, attempts to deposit to the target PCV deposit
+    function withdrawERC20RatioToSafeAddress(
+        address pcvDeposit,
+        address safeAddress,
+        address token,
+        uint256 basisPoints,
+        bool depositAfter
+    ) external override beforeAndAfterWithdrawal(pcvDeposit, safeAddress, depositAfter) {
+        require(basisPoints <= Constants.BASIS_POINTS_GRANULARITY, "PCVGuardian: basisPoints too high");
+        uint256 amount = (IERC20(token).balanceOf(pcvDeposit) * basisPoints) / Constants.BASIS_POINTS_GRANULARITY;
+        require(amount != 0, "PCVGuardian: no value to withdraw");
 
         IPCVDeposit(pcvDeposit).withdrawERC20(token, safeAddress, amount);
+        emit PCVGuardianERC20Withdrawal(pcvDeposit, safeAddress, token, amount);
+    }
+
+    // --------------- Modifiers  -------------
+
+    /// @notice modifier to factorize the logic in all withdrawals :
+    /// - first, ensure the deposit to withdraw from is unpaused
+    /// - second, perform withdrawal
+    /// - third, re-pause deposit if it was paused
+    /// - finally, call pcvDeposit.deposit() if depositAfter = true
+    modifier beforeAndAfterWithdrawal(
+        address pcvDeposit,
+        address safeAddress,
+        bool depositAfter
+    ) {
+        {
+            // scoped in this modifier to prevent stack to deep errors & enforce consistent acl
+            ICore _core = core();
+            require(
+                _core.hasRole(TribeRoles.GUARDIAN, msg.sender) ||
+                    _core.hasRole(TribeRoles.PCV_SAFE_MOVER_ROLE, msg.sender) ||
+                    _core.hasRole(TribeRoles.GOVERNOR, msg.sender),
+                "UNAUTHORIZED"
+            );
+            require(isSafeAddress(safeAddress), "PCVGuardian: address not whitelisted");
+        }
+
+        bool paused = pcvDeposit._paused();
+        if (paused) {
+            pcvDeposit._unpause();
+        }
 
-        if (pauseAfter) {
+        _;
```

Comment:
I think the modifier is better because we want the paused stated read before withdrawal to be used after withdrawal (to re-pause), we can't do that without modifier